Update create_graph.py

seigodev opened the pull request:
ダミーノードを挿入する機能を追加しました。

seigodev committed:
Add stack class and Count class.

seigodev committed:
Add 2 functions for inserting dummy nodes.

Add find_over_2_level_edges() and cut_edge_and_insert_dummy().
find_over_2_level_edges(): 2階層以上はなれているエッジの取得、およびリストへのダミーノードの追加を行う。
cut_edge_and_insert_dummy(): エッジの切断およびダミーノードの作成を行う。

aabaa commented on demo/create_graph.py:
```diff
@@ -141,6 +186,65 @@ def assign_x_coordinate(node_list):
+"""
+
+
+def find_over_2_level_edges(node_list):
```
findという名前の関数内で要素変更を行うのはNG．
cut_edges_higher_than_1_recursively などの関数名に変更してください．

aabaa commented on demo/create_graph.py:
```diff
@@ -141,6 +186,65 @@ def assign_x_coordinate(node_list):
+        source, target = cut_edge_stack.pop()
+        dummy = cut_edges_and_insert_dummy(source, target)
+        node_list.append(dummy)
+        if dummy.y - target.y > 1:
```
source.yとdummy.yとの差は1として大丈夫なのでしょうか？
もしそれが保証されているならcut_edges_and_insert_dummy()のコメントに記載してください．

aabaa commented on demo/create_graph.py:
```diff
@@ -141,6 +186,65 @@ def assign_x_coordinate(node_list):
+
+
+@Count
+def cut_edges_and_insert_dummy(source, target):
```
エッジを切ったらダミーノードが入るのは当たり前なので，関数名はcut_edgeなどで良いでしょう．
複数のエッジを切るのではないため，edgesではないです．
通常，関数名にandが入る場合はその関数を２つ以上の関数に分割するべきサインだと考えてください．

aabaa commented on demo/create_graph.py:
```diff
@@ -141,6 +186,65 @@ def assign_x_coordinate(node_list):
+    cut_edge_stack = Stack()
+    for target_node in node_list:
+        for source_node in target_node.source_nodes:
+            if source_node.y - target_node.y > 1:
```
些細ですが，高さを取得する関数を作成したほうが良いでしょう．
また，sourceとtargetのどちらが上にあるのかを意識しながらコーディングするのは宜しくありません．
absを使いましょう．
```python
def calc_edge_height(node1, node2):
    return abs(node1.y - node2.y)
```

aabaa commented on demo/create_graph.py:
```diff
@@ -141,6 +186,65 @@ def assign_x_coordinate(node_list):
+    Return:
+        dummy: sourceとtargetの間に挿入したダミーノード。Nodeオブジェクト。
+    """
+    dummy_counter = cut_edges_and_insert_dummy.count
```
念の為，エラーチェックしておくべきです
```python
assert calc_edge_height(source, target) > 1
```

seigodev committed:
Fix requested changes.

・find_over_2_level_edges() -> cut_edges_higher_than_1() (193行)
・add calc_edge_height() (219-229行)
・cut_edges_and_insert_dummy -> cut_edge() (223行)
・cut_edge()の返り値dummyの説明文の追加 (241-250行)
・assert calc_edge_height(source, target) > 1の追加(252行)
